### DISCUSSION TITLE
Could not load type Microsoft.AspNetCore.Internal.EndpointRoutingApplicationBuilderExtensions from assembly Microsoft.AspNetCore.Routing Ver=3.1.7.0

### DISCUSSION THREAD

**Discussion by krpranay** [dotnet/aspnetcore#25805] [General]
I am upgrading my dotnet core 2.2 app to 3.1. I have multiple services and one dotnet core base library all initially developed in 2.2 AspDotNetCore.  I am following 2 steps process.

1. First covert my base library to Dotnet standard 2.1. Which I am able to do successfully with few reference changes. At least I am able to compile it as Dotnet Standard 2.1 library.
2.  Now I am referencing this base 2.1 dotnet standard library in other AspDotnet core app which I am upgrading from 2.2 dotnet core to 3.1 dotnet core. This step is also successful and able to compile it.

But now the problem comes when I am trying to run this. Here is the error I am getting:
 
`Could not load type 'Microsoft.AspNetCore.Internal.EndpointRoutingApplicationBuilderExtensions' from assembly 'Microsoft.AspNetCore.Routing, Version=3.1.7.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'.

Where exactly it's failing while calling adding my custom routing middleware in below core function name `UseMvcCoreServices` in below code:

` app .UseMvcCoreServices(Configuration, Logger, env).UseElasticClient(Configuration, Logger);`

I am calling this in `Configure` function within Startup.cs. I am not very sure what exactly I am missing.

Anyone having any idea about how to fix it?

**Comment by davidfowl** [comment]
ASP.NET Core 2.2 and 3.1 are incompatible here. You can see the migration guide here https://docs.microsoft.com/en-us/aspnet/core/migration/22-to-30?view=aspnetcore-3.1&tabs=visual-studio#migrate-libraries-via-multi-targeting
